src/Microsoft.AspNetCore.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -154,6 +154,93 @@ public static void MergeAttributes(this TagHelperOutput tagHelperOutput, TagBuil
             }
         }
 
+        /// <summary>
+        /// Adds the given <paramref name="classValue"/> to the <paramref name="tagHelperOutput"/>'s
+        /// <see cref="TagHelperOutput.Attributes"/>.
+        /// </summary>
+        /// <param name="tagHelperOutput">The <see cref="TagHelperOutput"/> this method extends.</param>
+        /// <param name="classValue">The class value to add.</param>
+        public static void AddClass(
+            this TagHelperOutput tagHelperOutput,
+            string classValue)
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            if (!tagHelperOutput.Attributes.ContainsName("class"))
+            {
+                tagHelperOutput.Attributes.Add("class", classValue);
+            }
+            else
+            {
+                var classAttributeValue = tagHelperOutput.Attributes["class"].Value.ToString();
```

Comment:
sadly this isn't sufficient. Calling ToString on the value isn't sufficient in supporting all the various value types. @DamianEdwards played around with the idea of a class manager piece for TagHelperOutput. @DamianEdwards I tried looking for your code but couldn't find anything. Could you link so we could compare?

Reply:
@dougbu Will `ClassAttributeHtmlContent` cover all needed string cases?

Reply:
>  Will `ClassAttributeHtmlContent` cover all needed string cases?

Yes though it'll be much more efficient if you give up on the duplicate-avoiding part and unconditionally add the new `classValue`. The duplication doesn't actually change things.

That said, `ClassAttributeHtmlContent` doesn't help when improving the efficiency of the `RemoveClass()` method. But, that method will likely rarely be used. Tag helpers are generally additive.

Reply:
@dougbu Alright i will do those changes when I'm back from holiday next week. Thanks for the feedback

Reply:
@dougbu I really don't like the fact that the AddClass method can add multiples with the use of the `ClassAttributeHtmlContent` class. Can I at least create an extra extension called HasClass so users can check on their own if this class is already there?

Reply:
@Elfocrash Why are you concerned about duplicate classes? What underlying fact don't you like?

Reply:
@dougbu Generating "ugly" markup? I know it is not affecting functionality and if you have no problem with that then I'm fine.

Reply:
I would expect any helpers for managing class names to de-dupe them. Having class names appear more than once is ugly.

Reply:
:ok: "ugly" (as well as minification) is enough. We then need `HasClass()` and it must be efficient i.e. traverse the `Value` tree without `ToString()` unless you reach an `object`. In particular, it must handle `string`, `IHtmlContentContainer` and `IHtmlContent` efficiently.

For inspiration on the tree walks, look at `IHtmlContentBuilder` implementations in Razor or MVC.  Could for example write `HasClass()` using an `IHtmlContentBuilder` implementation with a `CopyTo(...)` method that only tracks whether the requested addition exists already.

A `TextWriter` similar to [`HasContentTextWriter`](https://github.com/aspnet/Mvc/blob/cfc05104e416f253bdfc07603f3e387c77aea975/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/DefaultEditorTemplates.cs#L446-L472) would also work though the semantics aren't as clean as using an `IHtmlContentBuilder`.

Reply:
Oops, you don't need the full `IHtmlContentBuilder`. `IHtmlContentContainer` is enough. I've updated my comment&hellip;

Reply:
Ok how about I keep the current de-duping code and i change the value to use `ClassAttributeHtmlContent` for now? We can expand on it with a `HasClass()` on it later if it proves problematic right?

Reply:
This conversation started with our objections to `Value.ToString()`. The `ToString()` in itself is problematic. If you can avoid that with the approach you describe, great.